### ISSUE TITLE
Qwen3.5-35B-A3B-GPTQ-Int4 Bringup

### ISSUE THREAD

**Issue by jtuyls** [iree-org/onnxruntime-ep-iree#51]
[qwen35b_full.zip](https://github.com/user-attachments/files/27448420/qwen35b_full.zip)

## State

Functional 40-layer Qwen3.5-35B-A3B-GPTQ-Int4 ONNX export, end-to-end through the IREE `onnxruntime` EP on AMD hip (gfx1100, RX 7900 XTX).

| Stage                       | Result                                  |
| --------------------------- | --------------------------------------- |
| Build                       | 40 layers, ~21.6 GB external weights    |
| Compile + parameter load    | ~23 s                                   |
| Decode inference (seq = 1)  | ~2.5 s                                  |
| Output                      | `(1, 1, 248320)` fp16 logits            |

The graph is hand-built with `onnx.helper`. **Almost everything is plain ONNX ops** that the EP routes through standard IREE codegen; the MoE expert dispatch is the only extern call. Composition:

- **10 SDPA layers** (indices `[3, 7, 11, 15, 19, 23, 27, 31, 35, 39]`) — pure ONNX. Q/K/V projections (MatMul), partial RoPE on the first 64 dims (Split/Neg/Concat/Mul/Add over Gather'd cos/sin tables), Q/K RMSNorm with `(1+w)*norm` style (Mul/ReduceMean/Add/Reciprocal/Sqrt/Mul), GQA expand 2 → 16 (Reshape/Tile), scaled MatMul/Softmax/MatMul, qgate sigmoid (Sigmoid/Mul), O projection (MatMul). KV-cache scatter via `Where` over a flat ring-buffer.
- **30 GDN linear-attention layers** — pure ONNX. QKV/Z/B/A projections, conv1d state shift+append (Slice/Concat), depthwise conv (Mul/ReduceSum/Sigmoid/Mul), L2 normalize (ReduceSum/Sqrt/Div), decay-by-`exp(g)`, rank-1 delta-rule update (Mul/ReduceSum/Sub/Mul/Add), gated RMSNorm + output projection. Recurrent + conv state stitched back into the flat buffers via Slice + Concat.
- **40 MoE blocks** — 4 `com.iree:ExternDispatch` nodes per call routing into the prebuilt expert kernel object (`kernels/moe_split_dispatch_qwen35-gfx1100.co`). **The only non-pure-ONNX part of the graph.**
- **Shared expert + embedding + final RMSNorm + lm_head** — pure ONNX (MatMul / Sigmoid / Mul / Gather).

So: 39 of every 40 layers' ops, plus all top/tail ops, go through the EP's standard ONNX → MLIR → IREE codegen path. Only the MoE expert dispatch in each layer is an extern HIP kernel.

## Layout

```
qwen35b_full/
├── attention_blocks.py            # SDPA + GDN ONNX helpers
├── moe_extern.py                  # 4-dispatch MoE ExternDispatch helper (decode)
├── moe_dispatch_gptq.mlir          # pre-generated MoE wrapper MLIR (referenced by EP)
├── kernels/                        # prebuilt MoE .co kernel objects
│   ├── moe_split_dispatch_qwen35-gfx1100.co
│   └── moe_wmma_prefill_qwen35-gfx1100.co
├── export_qwen35_real.py          # full multi-layer real-attention exporter
├── export_qwen35_moe_block.py     # single-layer sanity exporter
├── test_sdpa_only.py              # standalone SDPA helper test
├── test_gdn_only.py               # standalone GDN helper test
└── out/qwen35_real.onnx           # produced graph
    out/qwen35_real.weights.bin    # produced external weights (~21.6 GB)
```

## How to build / compile / run

Dependencies (in a fresh venv):

```bash
pip install \
    onnx onnxruntime numpy huggingface_hub safetensors torch \
    iree-base-compiler iree-base-runtime
# Plus the ort IREE EP shared library (this repo's build).
```

Build the ONNX (downloads weights from HF on first run, ~21.6 GB external bin):

```bash
python export_qwen35_real.py --num-layers 40 --output out/qwen35_real.onnx
```

Compile + run via the EP, single decode step:

```bash
python export_qwen35_real.py --num-layers 40 --run --device hip --target-arch gfx1100
```

Smaller smoke tests:

```bash
python test_sdpa_only.py --device hip       # 1-layer SDPA, random weights
python test_gdn_only.py  --device hip       # 1-layer GDN, random weights
python export_qwen35_real.py --num-layers 4 --run --device hip   # 4-layer e2e
```

The `--run` flag also dumps EP intermediates (compiled VMFB + IRPA) under `/tmp/iree_ep_*` for debugging.

### 1. Numerical correctness

The pipeline runs end-to-end but synthetic empty-state inputs (zero KV cache, zero recurrent / conv state, zero `seq_len`) produce all-NaN logits. This appears to be a model-level fp16 overflow — likely RMSNorm-of-zero-variance in the first attention layer or softmax over an all-masked SDPA window. Same behaviour observed in adjacent forward-pass implementations of this model.

Open questions:

- Where in the forward pass does NaN first appear? Bisect via `save_intermediates="1"`.
- Does it go away with a real tokenized prompt + a properly prefilled cache, or does it survive into real-prompt territory too?
- If the fp16 overflow is structural, what's the smallest fix — variance epsilon clamp, partial fp32 promotion, weight scaling?
- Is there a reference forward pass we can diff against layer-by-layer?

Concrete next step: write a real-prompt harness — HF tokenizer + short prompt + prefill (manually or by chunked decode) + decode loop + comparison to a reference implementation.

### 2. Prefill with dynamic sequence length

The exporter currently bakes `seq = 1` into every shape (decode-only) because the EP's `ExternDispatch` lowering doesn't emit `tensor<?xT>{%size}` annotations for dynamic outputs — see `src/mlir_gen.cc::EmitExternDispatch`.

There are also two model-level prefill gaps:

- **MoE prefill block.** A 5-dispatch WMMA-based variant exists in the kernel object (`moe_wmma_prefill_qwen35-gfx1100.co`); the ONNX side only wires the 4-dispatch decode path so far.
- **SDPA prefill.** Mechanical extension of the existing helper: causal mask becomes 2-D and the cache scatter writes a contiguous range. Should be a small change.
- **GDN prefill.** Per-token recurrence — needs either an ONNX `Loop` op or unrolling for a fixed prefill block size. The harder of the three.

Open questions:

- Should the EP support dynamic-output ExternDispatch (the "right" fix), or do we ship one ONNX file per supported prefill block size (e.g. 128, 256)?
- Is the GDN recurrence in a `Loop` op acceptable for the codegen backend, or do we want unrolled blocks?
- For prefill memory, is the existing flat KV-cache layout fine, or do we want a true paged layout to match other tooling?

### 3. Optimisation

Correctness first; once 1 + 2 are landed, several easy perf wins:

- **Fuse RMSNorm.** Each call site is currently 6+ ONNX ops (Mul/ReduceMean/Add/Reciprocal/Sqrt/Mul/Mul). ONNX 17+ has a built-in `LayerNormalization` op; switching cuts the dispatch count meaningfully (RMSNorm shows up ~80× across the graph).
- **Reduce Reshape/Transpose churn.** The helpers emit lots of defensive shape ops; constant folding handles most of them but the graph would be cleaner with a manual pass.
- **Multi-batch.** `bsz = 1` is baked in everywhere — generalising is mechanical.
- **Q/K norm `(1+w)`.** Currently emitted as `Add(w, 1) * x`, with the `Add` failing CPU constant folding (warns at session creation but the EP folds it). Pre-bake `w + 1` into the initializer.
- **Profile first.** Once correctness lands, profile a real-prompt decode step and let the data drive the next round.

Open questions:

- What's an acceptable ms/decode-token target for this bundle?
- How much of the slowness is MoE extern dispatch vs the 30 GDN layers' large dispatch count?
- Is graph-level fusion (RMSNorm, RoPE, gate*silu) worth it, or does the EP/IREE codegen already handle it?
